### ISSUE TITLE
Adopt GitFlow

### ISSUE THREAD

**Issue by blairconrad** [FakeItEasy/FakeItEasy#955]
This arose from a [Gitter conversation about how we support older releases](https://gitter.im/FakeItEasy?at=586104e2c895451b7520c61a).

@adamralph proposed moving from GitHub Flow to [GitFlow](http://nvie.com/posts/a-successful-git-branching-model/#hotfix-branches). @thomaslevesque and @blairconrad agree.

We'll need to  (at least - augment, since I'm sure I've forgotten things)

- [x] create the `develop` branch (it will be cleaner to wait until we have a new tag, such as 3.0.0-beta003. I'd consider just renaming master and making a new master off 3.0.0-beta002, but I think this would disrupt people)
- [x] protect the branch in GitHub
- [x] configure AppVeyor to make it the default branch
- [ ] update all our "contributing/how to build" documentation

Those are the obvious steps, I think.

Less obvious, I think we'll need to overhaul our "next release" workflow. It's been a little problematic anyhow, I think, since we've been guessing about our next release version and then renaming or replacing milestones and releases and build issues. Let's see how we can improve things, especially now that GitFlow has more to say about releases than does GitHubFlow.
I see two potential paths forward. I'm sure there are more.

**The first** is that we can work essentially as we've been doing, except not assign versions to milestones, releases, and release issues. We could make placeholders for "next release" and assign specific version names when we need them (when we create a release branch). When we create a hotfix branch, we will also create a new set of these things. In that case, they can be tied to a version right away.

**The second** is more radical, and maybe should be split off into another discussion, but I'll introduce it here. My biases abound.

* I generally find milestones to be nothing but overhead. We work by marking an issue in-progress, and then when it's done, we add it to the release. Once we release, the milestone provides no value at all. I'd consider ditching milestones. We could reserve them for special cases, like "breaking issues to go in the next major" or "things that have to go in 4.0, even though that release isn't on the horizon (like removing Self-Initializing Fakes)", but I think we could get along without them just about all the time
* I've long thought we should keep release notes in an `.md` file in the repo. As soon as a PR that's releasenotable is merged (or as part of the PR, if the contributor is savvy), we update the file with the release note (and thanks to the contributors). If we do this, we could make our "release" step generate the GitHub release from that file. We could do this manually at first, but I've been toying with the idea of a tool that'll scrape the file and create the GitHub Release. (Longer term, _I'd_ like to see this file be the source of our current version number, but that's slightly more involved as well and, I imagine, contentious.)
* Then all we need is a release checklist. We could either always have one around and just name it when we decide to target a particular version (i.e. when we create the release branch) **or** not even make one until we create the release or hotfix branch (with a script/build target that will create the branch and checklist for us).

**Comment by thomaslevesque** [comment]
> The first is that we can work essentially as we've been doing, except not assign versions to milestones, releases, and release issues. We could make placeholders for "next release" and assign labels when we need them (when we create a release branch). When we create a hotfix branch, we will also create a new set of these things. In that case, they can be tied to a version right away.

I like this option; it's easy and not too disruptive.

> I generally find milestones to be nothing but overhead. We work by marking an issue in-progress, and then when it's done, we add it to the release. Once we release, the milestone provides no value at all. I'd consider ditching milestones. We could reserve them for special cases, like "breaking issues to go in the next major" or "things that have to go in 4.0, even though that release isn't on the horizon (like removing Self-Initializing Fakes)", but I think we could get along without them just about all the time

I kind of like milestones. It makes it easier to see where we're headed, what goes in the next release, etc

> I've long thought we should keep release notes in an .md file in the repo. As soon as a PR that's releasenotable is merged (or as part of the PR, if the contributor is savvy), we update the file with the release note (and thanks to the contributors). If we do this, we could make our "release" step generate the GitHub release from that file. We could do this manually at first, but I've been toying with the idea of a tool that'll scrape the file and create the GitHub Release. (Longer term, I'd like to see this file be the source of our current version number, but that's slightly more involved as well and, I imagine, contentious.)

It's an interesting idea. However, if this file is going to contain the whole history of release notes, parsing it won't be trivial.

For now, I'd vote for the first option. We can always take some of your other ideas later if we want.

**Comment by blairconrad** [comment]
@thomaslevesque, thanks for the feedback. I agree that status quo (or near status quo) is a good first step. Other changes can always be considered later.

----

As an aside, 

> It's an interesting idea. However, if this file is going to contain the whole history of release notes, parsing it won't be trivial.

Actually, not so much, at least as I imagine it. Consider:

```markdown

## 3.0.0-beta003

### Changed
* Hide `object` members on several fluent API interfaces (#944)

### New
* Add a `ThrowsAsync` method to return a failed task (#932)

### Fixed
*  `Ignored`/`That` argument constraints do not Work for ByRef-Arguments in VisualBasic (#899)

### With special thanks for contributions to this release from:
* @thunderbird55

## 3.0.0-beta002

### New
* Automatic cancellation support #904

    Methods that accept a `CancellationToken` will now automatically throw an `OperationCanceledException` (or return a canceled task, for async methods) when they receive a canceled token.

…
```

You just read down to the previous `##` and copy it into the release.

**Comment by thomaslevesque** [comment]
> You just read down to the previous ## and copy it into the release.

True. And anyway, we could always use a Markdown parser if needed

**Comment by adamralph** [comment]
The milestones/release notes proposal has value but I do think it should be addressed in another issue.

I think it could be quite confusing (not only for us but for the rest of the community) if we switch to gitflow before 3.0.0, and I'm not sure it will give us much value before then. If we were using gitflow now, then what is currently `master` would be `develop` and what is currently `support-2` would be `master`. If we switch to gitflow right now (or after the next 3.0.0 beta) then we'll have `develop`, `master` and `support-2` with only `develop` being used correctly.

**Comment by blairconrad** [comment]
> If we were using gitflow now, then what is currently `master` would be `develop` and what is currently `support-2` would be `master`.

I'd've thunk that `master` would be `develop` and what is currently `3.0.0-beta002` would be `master`.
`support-2` wouldn't enter into it, I wouldn't think, as GitFlow doesn't address hotfixing releases that didn't come from `master`. It seems to assume that we'd want at most one hotfix off any release (of course, we're complicating things by considering adopting GitFlow for an existing project, but this limitation would exist regardless).

I'm guessing that your statement that `support-2` would be `master` implies a lower status for the pre-release versions, so they're released off `develop` instead of `master`? I fear we have different stances here, as I'd consider the pre-releases to be "releases" and think that even they would come off `master`.

**Comment by adamralph** [comment]
With gitflow, you only merge to master when you release stable.

**Comment by adamralph** [comment]
> I'm guessing that your statement that support-2 would be master implies a lower status for the pre-release versions, so they're released off develop instead of master?

That's correct. Typically, alphas and betas are released from develop, RC's from release-* and stables from master.

**Comment by blairconrad** [comment]
> With gitflow, you only merge to master when you release stable.
> That's correct. Typically, alphas and betas are released from develop, RC's from release-* and stables from master.

Hmm. Okay. That's different from my reading, but there's so much opportunity for interpretation. I'm still of the opinion that anything we'd want to put out on NuGet is a "release", even if the version contains a "-", but I'm happy to try out what you're used to.

On another note, how attached are you to the branch names as given in [A successful Git branching model](http://nvie.com/posts/a-successful-git-branching-model)? I've heard of variants that people use. In particular, some flavours name what Driessen terms "develop" as "master", with Driessen's "master" being called something else (e.g. "release", although alternatives such as "production" could also work). Since people are used to working off "master" when cloning GitHub projects, I can see it lowering the barrier to entry.
And this didn't occur to me at first, but then your comment 

>  If we were using gitflow now, then what is currently `master` would be `develop` and what is currently `support-2` would be `master`

becomes

>  If we were using gitflow now, then what is currently `master` would be `master and what is currently `support-2` would be `release`

(or another term for `release`)

And such a transformation would be quite easy.

**Comment by adamralph** [comment]
I'd probably prefer to stick to the usual branch names and definitions, as defined in that article. What we are currently calling `support-2` is what `master` is for in gitflow, and what we are currently calling `master` is what `develop` is for in gitflow. `release-` branches only exist for a short time, in order to "stabilise" `develop`. Indeed, in our workflow, I beilieve the only time we'd only require a `release` branch is when we want to release an RC.

For me, the natural way for us to switch to gitflow would be:

- release 3.0.0 RTM from `master`
- create a `develop` branch from the head of `master`

From that point on we can follow gitflow as described in that article.

**Comment by thomaslevesque** [comment]
I'd also prefer to use the standard branch names from gitflow.

> Since people are used to working off "master" when cloning GitHub projects, I can see it lowering the barrier to entry.

I don't think it really matters; you can change the default branch in Github, and it will automatically be checked out when you clone the repo.

> For me, the natural way for us to switch to gitflow would be:
> 
> - release 3.0.0 RTM from master
> - create a develop branch from the head of master

Sounds good to me

**Comment by adamralph** [comment]
The repos I currently work on which use gitflow do indeed have their default branch set to develop. It doesn't seem to cause any problems.

**Comment by blairconrad** [comment]
Do we now have to wait to do this until after we release 3.1.0 (or 3.0.1)? `master` has already moved beyond 3.0.0, so it might be confusing to switch back.

**Comment by adamralph** [comment]
Yes, I think it would be better to wait until the tip of master is released (3.10 or 3.0.1) and then create a `develop` branch.

**Comment by thomaslevesque** [comment]
We can wait. On the other hand, the only commits made to `master` since RC1 are non-functional changes, so it wouldn't really matter if we released 3.0.0 from the tip of `master` (we would have to revert the commit that bumps the version to 3.1.0, of course)

**Comment by thomaslevesque** [comment]
Fixed in [release 3.3.1](https://github.com/FakeItEasy/FakeItEasy/releases/tag/3.3.1)
